City-Bureau/city-scrapers-wichita
# Feat: upload board docs agendas

## What's this PR do?
Adds a new feature to the boarddocs mixin that uploads agendas to Azure storage.

## Why are we doing this?
Meeting materials can only be accessed using a POST request so we aren't able to provide a list of links in the traditional way. Instead, we first upload meeting materials to a city-scraper Azure container and then return a list of links to those materials in the spider's output.

## Steps to manually test

Several spiders use the boarddocs mixin. I suggest using `wicks_andover_boe`:

1. Ensure the project is installed:
```
pipenv sync --dev
```
2. Activate the virtual env and enter the pipenv shell:
```
pipenv shell
```
3. Go into our Azure account for the cityscraper project and delete any existing agendas that exist in the `meetings-downloads` container.

4. Run the spider in prod, providing Azure account_name and account_key:
```
scrapy crawl wicks_andover_boe -s CITY_SCRAPERS_ENV=prod -s AZURE_ACCOUNT_NAME=<account_name> -s AZURE_ACCOUNT_KEY=<account_key>
```
5. Monitor the output and ensure no errors are raised.

6. Inspect `test_output.csv` to ensure the data looks valid.

7. Access the agendas in `links` field to make sure they download as expected. The URL should be to our Azure container.

## Are there any smells or added technical debt to note?
- It would probably be ideal to have some more test coverage for this new behavior but I'll admit I was finding it tricky to properly mock the mixin's new `upload_agenda` method.
- The agenda files are in HTML. I considered converting them to PDFs before uploading them but the python package I used was causing me headaches and a number of other conversion libraries seemed to require installing binaries. I think for most folks reading these agendas, the HTML files should be easy enough to open and peruse. It's simple markup with no JS or other assets (as far as I'm aware).

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1207425189153432
